src/util/dirty.cpp

```diff
@@ -13,47 +17,166 @@
+    {
+        nPages = faabric::util::getRequiredHostPages(region.size());
+        dirtyFlags = std::vector<char>(nPages, 0);
+        regionBase = region.data();
+        regionTop = region.data() + region.size();
+    }
+
+    virtual void markPage(void* addr)
+    {
+        long pageNum = ((uint8_t*)addr - regionBase) / HOST_PAGE_SIZE;
+        dirtyFlags[pageNum] = 1;
+    }
+
+    virtual bool isInitialised() { return regionTop != nullptr; }
+
+    virtual int getNPages() { return nPages; }
+
+    virtual std::vector<char> getDirtyFlags() { return dirtyFlags; }
+
+    virtual void reset()
+    {
+        nPages = 0;
+        dirtyFlags.clear();
+        regionBase = nullptr;
+        regionTop = nullptr;
+    }
+
+  protected:
+    int nPages = 0;
+
+    // std::vector<bool> here seems to worsen performance by >4x
+    // std::vector<char> seems to be optimal
+    std::vector<char> dirtyFlags;
+
+    uint8_t* regionBase = nullptr;
+    uint8_t* regionTop = nullptr;
 };
 
-DirtyTracker& getDirtyTracker()
+/**
+ * Thread-safe wrapper around dirty tracking data, necessary for use with the
+ * background event thread in the userfaultfd tracker. Although the logic around
+ * the interaction between the event thread and the main thread should in theory
+ * protect against concurrent accesses, we need this to keep TSan happy.
+ */
+class ThreadsafeDirtyTrackingRecord final : public DirtyTrackingRecord
 {
-    static DirtyTrackerSingleton dt;
+  public:
+    ThreadsafeDirtyTrackingRecord() = default;
+
+    void trackRegion(std::span<uint8_t> region) override
+    {
+        FullLock lock(mx);
+        DirtyTrackingRecord::trackRegion(region);
+    }
 
-    std::string trackMode = faabric::util::getSystemConfig().dirtyTrackingMode;
-    if (trackMode == "softpte") {
-        return dt.softpte;
+    void markPage(void* addr) override
+    {
+        FullLock lock(mx);
+        DirtyTrackingRecord::markPage(addr);
     }
 
-    if (trackMode == "segfault") {
-        return dt.sigseg;
+    std::vector<char> getDirtyFlags() override
+    {
+        SharedLock lock(mx);
+        return DirtyTrackingRecord::getDirtyFlags();
     }
 
-    if (trackMode == "none") {
-        return dt.none;
+    bool isInitialised() override
+    {
+        SharedLock lock(mx);
+        return DirtyTrackingRecord::isInitialised();
     }
 
-    throw std::runtime_error("Unrecognised dirty tracking mode");
+    int getNPages() override
+    {
+        SharedLock lock(mx);
+        return DirtyTrackingRecord::getNPages();
+    }
+
+    void reset() override
+    {
+        FullLock lock(mx);
+        DirtyTrackingRecord::reset();
+    }
+
+  private:
+    std::shared_mutex mx;
+};
+
+void* pageAlignAddress(void* faultAddr)
+{
+    uintptr_t addr = (uintptr_t)faultAddr;
+    addr &= -HOST_PAGE_SIZE;
+    auto* alignedAddr = (void*)addr;
+    return alignedAddr;
+}
+
+// Thread-local tracking information for dirty tracking using signal
```

Comment:
I think the style is consistent, just not in the way you're thinking. I use the following convention:

- Class and method comments that will go into documentation (i.e. above class and method declarations) with `/** */`
- All other single or multi-line comments with `//`

Now that I think about it, this is actually Javadoc style that I picked up years ago, and probably not in line with the Moz C/C++ style guide. I've used it in more than just this PR, so if we want to change it, we should do a codebase-wide effort and ideally enforce it in the formatting checks.

Reply:
Given that we're using Doxygen we should go with what they say in this case rather than the C/C++ guide: https://www.doxygen.nl/manual/docblocks.html.

Their first example uses Javadoc style as I mention, so I'd be happy to stick with that. It makes it easier to distinguish between class/ method docs (`/**`) and implementation-specific comments (i.e. `//`).

Let's discuss offline though.